File: scripts/create_user.py
#!/usr/bin/env python3
"""Создать или обновить пользователя Timesheet."""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from werkzeug.security import generate_password_hash

from server import app, get_db, init_db


def main() -> int:
    parser = argparse.ArgumentParser(description="Create or update a Timesheet user")
    parser.add_argument("username")
    parser.add_argument("password")
    parser.add_argument("--role", choices=("user", "admin"), default="user")
    parser.add_argument("--fio", default="", help="default_fio for user role")
    args = parser.parse_args()

    with app.app_context():
        init_db()
        db = get_db()
        password_hash = generate_password_hash(args.password)
        existing = db.execute(
            "SELECT id FROM users WHERE username = ?", (args.username,)
        ).fetchone()
        if existing:
            db.execute(
                "UPDATE users SET password_hash = ?, role = ?, default_fio = ? WHERE id = ?",
                (password_hash, args.role, args.fio, existing["id"]),
            )
            action = "updated"
        else:
            db.execute(
                "INSERT INTO users (username, password_hash, role, default_fio) VALUES (?, ?, ?, ?)",
                (args.username, password_hash, args.role, args.fio),
            )
            action = "created"
        db.commit()

    print(f"User {args.username!r} {action} (role={args.role})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/create-user.js
#!/usr/bin/env node
'use strict';

// Создать или обновить пользователя Timesheet.
// Usage: node scripts/create-user.js <username> <password> [--role user|admin] [--fio "ФИО"]

const path = require('path');
const { db, initDb, generatePasswordHash } = require(path.join(__dirname, '..', 'server.js'));

function parseArgs(argv) {
  const positional = [];
  const options = { role: 'user', fio: '' };
  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === '--role') {
      options.role = argv[++i];
    } else if (arg === '--fio') {
      options.fio = argv[++i];
    } else {
      positional.push(arg);
    }
  }
  return { positional, options };
}

function main() {
  const { positional, options } = parseArgs(process.argv.slice(2));
  const [username, password] = positional;

  if (!username || !password) {
    console.error('Usage: node scripts/create-user.js <username> <password> [--role user|admin] [--fio "ФИО"]');
    return 1;
  }
  if (!['user', 'admin'].includes(options.role)) {
    console.error(`Invalid --role: ${options.role} (expected user|admin)`);
    return 1;
  }

  initDb();
  const passwordHash = generatePasswordHash(password);
  const existing = db.prepare('SELECT id FROM users WHERE username = ?').get(username);
  let action;
  if (existing) {
    db.prepare('UPDATE users SET password_hash = ?, role = ?, default_fio = ? WHERE id = ?').run(
      passwordHash,
      options.role,
      options.fio,
      existing.id
    );
    action = 'updated';
  } else {
    db.prepare('INSERT INTO users (username, password_hash, role, default_fio) VALUES (?, ?, ?, ?)').run(
      username,
      passwordHash,
      options.role,
      options.fio
    );
    action = 'created';
  }

  console.log(`User '${username}' ${action} (role=${options.role})`);
  return 0;
}

process.exitCode = main();
